TASK: Verify and optimize FlowQuery operations

mficzel opened the pull request:
Upmerge the flowQuery Tests from 8.3, fix bugs and optimize the implementation to reduce the amount of db-queries as far as possible without introducing breakiness. The tests for `unique` and `remove` are added here because those operations did not exist in Neos 8.3.

The optimizations had to be very defensive and could not improve the `children` and `find` much since the following fizzle features could not be supported by PropertyValueConstraints: 

- Case insensitive comparisons `=~`, `^=~`, `*=~` and `$=~` (this may be fixed with #4721)
- Comparison of nested properties life `foo.bar = "baz"`
- Comparison of array entities in `^=`, `*=` and `$=`

This may be optimized further in a future release or a third party package that allows itself more breakiness. 

Notes about changes compared to Neos 8.3: 
- The order of nodes returned by `find` with typeFilter has changed. Since this was never guaranteed this is ok

In addition:
- `Nodes::last` method is added
- `Nodes::until` is removed because of broken semantic. BetterAlternatives are `previousAll` and `nextAll`
- `Nodes::isEmpty` add assertions that `first` and `last` will return not null
- `NodeNameFactory::forRoot` is added

Relates: #4702, #4695

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

This pr contains the upmerge of #4695 and #4702 (follow up) as this will otherwise cause lots of errors after upmerging. 
You can compare this by checking out this pr and compare the content of `Neos.Neos/Tests/Behavior/Features/Fusion/FlowQuery.feature` with the 8.3 branch. 

I consider the current behavior (with this pr) ok for a beta 1 as find (with nodetypes) already returns the correct nodes but ordered differently.


**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mficzel committed:
Merge branch 'task/addBehatTestsForFlowQueryCrOperations' into 90/verifyAndOptimizeFlowQuery

# Conflicts:
#	Neos.Neos/Documentation/References/CommandReference.rst
#	Neos.Neos/Documentation/References/EelHelpersReference.rst
#	Neos.Neos/Documentation/References/FlowQueryOperationReference.rst
#	Neos.Neos/Documentation/References/Signals/ContentRepository.rst
#	Neos.Neos/Documentation/References/Signals/Flow.rst
#	Neos.Neos/Documentation/References/Signals/Media.rst
#	Neos.Neos/Documentation/References/Signals/Neos.rst
#	Neos.Neos/Documentation/References/Validators/Flow.rst
#	Neos.Neos/Documentation/References/Validators/Media.rst
#	Neos.Neos/Documentation/References/Validators/Party.rst
#	Neos.Neos/Documentation/References/ViewHelpers/ContentRepository.rst
#	Neos.Neos/Documentation/References/ViewHelpers/FluidAdaptor.rst
#	Neos.Neos/Documentation/References/ViewHelpers/Form.rst
#	Neos.Neos/Documentation/References/ViewHelpers/Fusion.rst
#	Neos.Neos/Documentation/References/ViewHelpers/Media.rst
#	Neos.Neos/Documentation/References/ViewHelpers/Neos.rst
#	Neos.Neos/Documentation/References/ViewHelpers/TYPO3Fluid.rst

mficzel committed:
TASK: Adjust test for the new cr

mficzel committed:
TASK: Fix ParentsUntil Operation and add tests for `unique` and `remove`

mficzel committed:
Merge branch 'task/addBehatTestsForFlowQueryCrOperations' into 90/verifyAndOptimizeFlowQuery

# Conflicts:
#	Neos.Neos/Tests/Behavior/Features/Fusion/FlowQuery.feature

mficzel committed:
TASK: Fix `NextUntil` and  `PrevUntil` and optimize `Next`, `NextAll`, `Prev`, `PrevAll`

In addition:
- `Nodes::last` method is added
- `Nodes::until` is removed because of broken semantic. BetterAlternatives are `previousAll` and `nextAll`
- `NodeNameFactory::forRoot` is added
- `Nodes::isEmpty` add assertions that `first` and `last` will return not null

mhsdesign commented:
Fyi to why `until` will be removed https://neos-project.slack.com/archives/C3MCBK6S2/p1699020664345879

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php:
```diff
@@ -116,6 +128,10 @@ public function reverse(): self
     }
 
+    /**
+     * @phpstan-assert-if-false !null $this->first()
+     * @phpstan-assert-if-false !null $this->last()
```
```suggestion
     * @phpstan-assert-if-true Node $this->first()
     * @phpstan-assert-if-true Node $this->last()
```

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php:
```diff
@@ -14,6 +14,8 @@
 
 namespace Neos\ContentRepository\Core\Projection\ContentGraph;
 
+use function Amp\Promise\first;
+
```
```suggestion
```

mhsdesign commented:
thanks for taking care. It looks quite logical and way more performant ^^. I like to wait for https://github.com/neos/neos-development-collection/pull/4695 to be upmerged so we easier review the tests.

mficzel committed:
Merge branch 'task/addBehatTestsForFlowQueryCrOperations' into 90/verifyAndOptimizeFlowQuery

# Conflicts:
#	Neos.Neos/Tests/Behavior/Features/Fusion/FlowQuery.feature

mficzel committed:
TASK: Disable the failing parts of `find` operation and adjust the oder expectation of type filter

mficzel committed:
Apply suggestions from code review

Co-authored-by: Marc Henry Schultz <85400359+mhsdesign@users.noreply.github.com>

mficzel commented:
@mhsdesign if you look into the commits you will find that this pr contains the upmerge of #4695 and #4702 (follow up) you can compare the test cases by checking out the pr and use compare branches on the feature file to see the differences.

mficzel committed:
TASK: Adjust format of the todo notice in `find` operation

mficzel commented:
I suggest to look into the `children` and `find` operation again in a follow pr as those are more complex and still look weird inside.

mficzel committed:
TASK: Fix phpstan type assertation annotations

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 90/verifyAndOptimizeFlowQuery

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/Fusion/FlowQuery.feature:
```diff
@@ -318,13 +349,44 @@ Feature: Tests for the "Neos.ContentRepository" Flow Query methods.
     """fusion
     test = Neos.Fusion:DataStructure {
+      # @todo Decide wether the changed order of the results compared to Neos 8.3 is ok
+      # Result in Neos 8.3: "typeFilter: a1a,a1b1a,a1a2,a1b2,a1a3,a1a4,a1a5,a1a6"
+      # Result in Neos 9.0: "typeFilter: a1a,a1a2,a1b2,a1a3,a1a4,a1a5,a1a6,a1b1a"
       typeFilter = ${q(node).find('[instanceof Neos.Neos:Test.DocumentType2]').get()}
-      combinedFilter = ${q(node).find('[instanceof Neos.Neos:Test.DocumentType2][uriPathSegment*="b1"]').get()}
+      # @todo Fix and re enable `combinedFilter` case
+      # Result in Neos 8.3: "combinedFilter: a1b1a"
+      # Result in Neos 9.0: "combinedFilter: a1a,a1a2,a1b2,a1a3,a1a4,a1a5,a1a6,a1b1a"
+      # combinedFilter = ${q(node).find('[instanceof Neos.Neos:Test.DocumentType2][uriPathSegment*="b1"]').get()}
+      @process.render = Neos.Neos:Test.RenderNodesDataStructure
+    }
```
Should we resolve this todo here or in a followup?

mficzel replied:
@mhsdesign in a follow up. I did no changes for those operations so this only documents the deviation from 8.3 to 9.0 that was there. The fixes for the other operations already optimize quite a bit that should go to beta1

mficzel replied:
`children` and `find` will have to wait for beta 2

mficzel replied:
Might also be that `filter` could benefit from this https://github.com/neos/neos-development-collection/pull/4701

mhsdesign replied:
okay i thought so already so lets not block this any further ;)

mhsdesign commented on Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/PrevUntilOperation.php:
```diff
@@ -70,22 +71,21 @@ public function evaluate(FlowQuery $flowQuery, array $arguments): void
-            if (isset($until[0]) && !empty($until[0])) {
-                $prevNodes = $prevNodes->until($until[0]);
+            if (isset($untilNodes) && !$untilNodes->isEmpty()) {
+                $prevNodes = $prevNodes->previousAll($untilNodes->first())->reverse();
```
i dont fully grasp this line, and in `NextAllUntil` we use

```
$nextNodes = $nextNodes->previousAll($untilNodes->first());
```

but it seems to work :D

mficzel replied:
Was needed to get the order we had before. Can check the priorities once more and upgrade this today. This can be done also after the merge.

mficzel replied:
I suggest to discuss about changing the order later. Since this does nothing bad it could be merged for beta 1.

mhsdesign reviewed (commented):
Thanks great work. Some operations have different priorities either 0 or 100 - probably just due to copy pasta at some point.

i also want to point out that the order of returned items doesnt seem to always reflect the cr philosophy of the closest first, see `PrevUntil` expected `a1a3,a1a2` actual `a1a2,a1a3` 


So maybe we should fix this - in another cleanup? - and advise people to just reverse if they need the old behaviour ...

mhsdesign reviewed (approved):
;)

mficzel committed:
TASK: Add sortNames and priority to flowQuery operations

mficzel commented:
@mhsdesign pushed a change to 
- restore the operation priority as it was in 8.3 (100 for operatins that override derfault flowQuery operations and 0 otherwise)
- define shortNames for all flowQuery operations

... pretty sure the complaints of the linter are not caused by me ...

mficzel commented:
Just seen that the same errors now pop up in the 9.0 branch. ... now i am sure i am not guilty

mhsdesign commented:
Thanks @mficzel now its perfect ;)

mficzel committed:
Merge branch '9.0' into 90/verifyAndOptimizeFlowQuery

mficzel committed:
TASK: Add special handling of absolute node pathes since fizzle does not like the new syntax (yet)

mficzel committed:
TASK: Fix find operation

Absolute pathes are handled separately as the fizzle parser cannot handle the syntax

mhsdesign committed:
TASK: Enable behat strict mode to fail tests without snippet implementation

mhsdesign commented:
I made behat test now run in strict mode, as previously a non implemented snipped would not cause behat to fail???
Cant do it on the 9.0 branch as the current 8.3 upmerged flowquery test would fail in that case 😂

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 90/verifyAndOptimizeFlowQuery

kitsunet reviewed (approved):
Very nice, thank youuuuuu! :)
